### PR TITLE
Impl simple mamba model

### PR DESCRIPTION
**Pull request by drbh**
This draft PR is a work in progress implementation of the mamba model. This PR currently loads weights, and produces correct logits after a single pass. 

This PR still needs to correctly integrate this model so it produces tokens as expected, and apply optimization to avoid all copies during runtime/unnecessary operations.

#### Helpful resources
[Mamba: Linear-Time Sequence Modeling with Selective State Spaces (Albert Gu and Tri Dao)](https://arxiv.org/abs/2312.00752)
https://github.com/johnma2006/mamba-minimal
https://github.com/huggingface/candle/blob/main/candle-examples/examples/mamba-minimal/model.rs
https://github.com/huggingface/transformers/pull/28094

Notes: this dev work is currently targeting `state-spaces/mamba-130m`, so if you want to test please use that model. Additionally when starting the router the prefill needs to be limited: `cargo run -- --max-batch-prefill-tokens 768 --max-input-length 768`

## Update / Current State

Integration tests have been added and basic functionality such as model loading is supported.

```bash
cd integration-tests
pytest -vv models/test_fused_kernel_mamba.py
```
- [x] add tests
- [x] load model
- [x] make simple request 
- [ ] resolve warmup issue
- [ ] resolve output issues

fetching models tested during dev
```bash
text-generation-server download-weights state-spaces/mamba-130m
text-generation-server download-weights state-spaces/mamba-1.4b
text-generation-server download-weights state-spaces/mamba-2.8b
```

The server can be run 
```bash
cd server
 MASTER_ADDR=127.0.0.1 MASTER_PORT=5555 python text_generation_server/cli.py serve state-spaces/mamba-2.8b
```

router
```bash
cargo run
```

make a request
```bash
curl -s localhost:3000/generate \
    -X POST \
    -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":20}}' \
    -H 'Content-Type: application/json' | jq
```

response
```json
{
  "generated_text": "\n\nDeep learning is a machine learning technique that uses a deep neural network to learn from data."
}
```

### CODE REVIEW

**Comment by RonanKMcGovern** [comment]
thanks, this will be a great addition as we see more mamba architectures

File: server/text_generation_server/models/__init__.py [hunk 1]
```diff
@@ -161,7 +162,25 @@ def get_model(
...
+            )
+
+    if model_type == "ssm":
+        return Mamba(
+            model_id,
+            revision,
+            quantize=quantize,
+            dtype=dtype,
+            trust_remote_code=trust_remote_code,
+        )
```
Comment (drbh) [inline server/text_generation_server/models/__init__.py:183]: currently the models hosted on the hub under https://huggingface.co/state-spaces do not include a `model_type` and look like

```json
{
    "d_model": 2560,
    "n_layer": 64,
    "vocab_size": 50277,
    "ssm_cfg": {},
    "rms_norm": true,
    "residual_in_fp32": true,
    "fused_add_norm": true,
    "pad_vocab_size_multiple": 8
}
```

once this is updated we can remove this logic

File: server/text_generation_server/models/custom_modeling/mamba_modeling.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            self.dt_proj_weight,
+            self.out_proj_weight,
+            self.out_proj_bias,
+            self.negA,
+            None,
+            None,
+            self.D.float(),
+            delta_bias=self.dt_proj_bias.float(),
+            delta_softplus=True,
+        )
```
Comment (drbh) [inline server/text_generation_server/models/custom_modeling/mamba_modeling.py:80]: would really like to avoid the `transpose` but it seems that the kernel is expecting these dimensions... any suggestions to avoid would be greatly appreciated!

File: server/text_generation_server/models/mamba.py [hunk 3]
```diff
@@ -287,8 +299,60 @@ def concatenate(cls, batches: List["MambaBatch"]) -> "MambaBatch":
...
+        for batch in batches:
+            for i in range(n_blocks):
+                conv_state, ssm_state = batch.inference_params.key_value_memory_dict[i]
+                batch_size = batch.inference_params.max_batch_size
+                try:
+                    inference_params.key_value_memory_dict[i][0][current_batch:current_batch + batch_size] = conv_state
+                    inference_params.key_value_memory_dict[i][1][current_batch:current_batch + batch_size] = ssm_state
+                except Exception:
+                    import ipdb;ipdb.set_trace()
+                    pass
```
Comment (drbh) [inline server/text_generation_server/models/mamba.py:352]:
```suggestion
```

**Review by Narsil — APPROVED** [review]
LGTM
